React/Testing best practices.md

```diff
@@ -0,0 +1,610 @@
+# React Testing
+
+Tools we used for testing are [React-Testing-Library](https://testing-library.com/docs/) and [Jest](https://jestjs.io/docs/en/getting-started)
+
+## Setup
+
+First of all install jest
+
+```bash
+npm install --save-dev jest @types/jest
+```
+
+Add test script to `package.json`
+
+```js
+  "scripts": {
+    "test": "jest",
+    ...
+  }
+```
+
+Create `jest.config.js` in root of project
+
+```js
+module.exports = {
+  testPathIgnorePatterns: ["<rootDir>/.next/", "<rootDir>/node_modules/"],
+  moduleDirectories: ["node_modules", "src"],
+};
+```
+
+Add `preset` to `.babelrc`
+
+```js
+{
+  "presets": [
+    [
+      "next/babel"
+    ]
+  ]
+}
+```
+
+Next install react-test-library
+
+```bash
+npm i --save-dev @testing-library/react
+```
+
+### User events
+
+For mocking events [testing-library/user-event](https://github.com/testing-library/user-event) is used.
+
+```bash
+npm install --save-dev @testing-library/user-event
+```
+
+### Testing hooks
+
+For testing hooks [react-hooks-testing-library](https://github.com/testing-library/react-hooks-testing-library) is used.
+
+```bash
+npm install --save-dev @testing-library/react-hooks
+```
+
+### Add svg support
+
+```bash
+npm install --save-dev jest-svg-transformer
+```
+
+Add `transform` section to `jest.config.js`
+
+```js
+module.exports = {
+  ...
+  transform: {
+    '^.+\\.tsx?$': 'babel-jest',
+    '^.+\\.svg$': 'jest-svg-transformer',
+  },
+}
+```
+
+## Utils
+
+### Mock providers and store
```

Comment:
I have one question for this section. What if some component needs some specific provider that is not needed anywhere else? Would you add it to this one file that is used for testing all components, or would you create another customRenderer that is based on this global customRenderer, while also providing mocks for those additional providers that are specific to the component that I am testing?

Reply:
In that case we can just wrap our component in that provider in that test.

Example:
```tsx
it("Is rendering", () => {
  const { getByText } = render(
    <SpecialProvider>
      <UserCard />
    </SpecialProvider>
  );

  expect(getByText(/username/i)).toBeDefined();
});
```

Not sure if that should be moved to a separate file because it's used only in this test. cc @isBatak